actors/builtin/miner/miner_actor.go

```diff
@@ -2067,6 +2068,253 @@ func (a Actor) RepayDebt(rt Runtime, _ *abi.EmptyValue) *abi.EmptyValue {
+			rt.Log(rtt.INFO, "failed to get sector, skipping sector %d", update.SectorID)
+			continue
+		}
+
+		// TODO: this will change to "was originally dealless, and is currently no active deals"
+		if len(sectorInfo.DealIDs) != 0 {
+			rt.Log(rtt.INFO, "cannot update sector with deals, skipping sector %d", update.SectorID)
+			continue
+		}
+
+		// TODO: confirm sector is healthy
+
+		// TODO: What's the impact of the following flow:
+		// 	- activating the deals
+		// - failing to verify the proof (or some other failure later)
+		// - allowing the message to succeed anyway ("skipping" the sector)
+		// Does that leave the deal uncommitable (since state doesn't get reverted)? is it better to just fail the msg, revert state, and allow for a retry later?
+		code := rt.Send(
+			builtin.StorageMarketActorAddr,
+			builtin.MethodsMarket.ActivateDeals,
+			&market.ActivateDealsParams{
+				DealIDs:      update.Deals,
+				SectorExpiry: sectorInfo.Expiration,
+			},
+			abi.NewTokenAmount(0),
+			&builtin.Discard{},
+		)
+
+		if code != exitcode.Ok {
+			rt.Log(rtt.INFO, "failed to activate deals on sector %d, skipping sector %d", update.SectorID)
+			continue
+		}
+
+		prevalidatedUpdates = append(prevalidatedUpdates, &updateAndSectorInfo{
+			update:     &update,
+			sectorInfo: sectorInfo,
+		})
+
+		sectorsDeals = append(sectorsDeals, market.SectorDeals{DealIDs: update.Deals, SectorExpiry: sectorInfo.Expiration})
+		sectorsDataSpec = append(sectorsDataSpec, &market.SectorDataSpec{
+			SectorType: sectorInfo.SealProof,
+			DealIDs:    update.Deals,
+		})
+	}
+
+	//// TODO: this method has some potential aborts -- are those all programmer error or is there scope to improve UX for bad input?
+	dealWeights := requestDealWeights(rt, sectorsDeals)
+
+	builtin.RequirePredicate(rt, len(dealWeights.Sectors) == len(prevalidatedUpdates), exitcode.ErrIllegalState,
+		"deal weight request returned %d records, expected %d", len(dealWeights.Sectors), len(prevalidatedUpdates))
+
+	// TODO: this method has some potential aborts -- are those all programmer error or is there scope to improve UX for bad input?
+	unsealedSectorCIDs := requestUnsealedSectorCIDs(rt, sectorsDataSpec...)
+
+	builtin.RequirePredicate(rt, len(unsealedSectorCIDs) == len(prevalidatedUpdates), exitcode.ErrIllegalState,
+		"unsealed sector cid request returned %d records, expected %d", len(unsealedSectorCIDs), len(prevalidatedUpdates))
+
+	succeededSectors := bitfield.New()
+	var st State
+	rt.StateTransaction(&st, func() {
+		for k, updateWithSectorInfo := range prevalidatedUpdates {
+			err = rt.VerifyReplicaUpdate(
+				proof.ReplicaUpdateInfo{
+					NewSealedSectorCID:   updateWithSectorInfo.update.NewSealedSectorCID,
+					OldSealedSectorCID:   updateWithSectorInfo.sectorInfo.SealedCID,
+					NewUnsealedSectorCID: unsealedSectorCIDs[k],
+				})
+
+			if err != nil {
+				rt.Log(rtt.INFO, "failed to verify replica proof, skipping sector %d", updateWithSectorInfo.sectorInfo.SectorNumber)
```

Comment:
As noted above this is going to leave the market actor in a bad state.  Your options are failing the whole message or moving this to above market actor side effects in the above loop.  The second option should be doable -- I don't think the VerifyReplicaUpdate and inputs require any state modification so you can take them out of the StateTransaction and put them above.

Reply:
The reason not to do that is that it requires the data commitments to be all calculated before the more basic per-update checks (and i assume that isn't cheap).

I do think the optimal thing to do is to make 3 per-update passes -- 1) quick checks, 2) expensive checks / operations (proof verif & deal activation), and 3) actual miner / power state changes. CommD calculation can then be sandwiched between 1 and 2.

Feels a bit bloated, though. Tell me whether you agree and I'll implement that.

Reply:
I'm more in favor of just one set of checks and then the atomic updates.  It's good to handle failures in batch gracefully but you're already meeting my expectations for that.  The most important failures to handle without killing the batch are the epoch dependent ones that can't be caught during gas estimation, and you are already doing deal activation this way.

Reply:
That makes a lot of sense, yeah. Sounds good!